File: key-value-store/implementation/python/test_kv_store.py
import unittest

from kv_store import DistributedKVStore


class TestDistributedKVStore(unittest.TestCase):
    def test_put_get_single_value(self):
        kv = DistributedKVStore(["s1", "s2", "s3"], n_replicas=3, w_quorum=2, r_quorum=2)
        kv.put("user:1", "alice", coordinator="s1")
        versions = kv.get("user:1")
        self.assertEqual(1, len(versions))
        self.assertEqual("alice", versions[0].value)

    def test_concurrent_writes_create_conflict_versions(self):
        kv = DistributedKVStore(["s1", "s2", "s3"], n_replicas=3, w_quorum=2, r_quorum=2)

        base = kv.put("profile:name", "john", coordinator="s1")
        kv.put("profile:name", "johnSF", coordinator="s2", base_clock=base)
        kv.put("profile:name", "johnNY", coordinator="s3", base_clock=base)

        versions = kv.get("profile:name")
        values = sorted(v.value for v in versions)
        self.assertEqual(["johnNY", "johnSF"], values)

    def test_conflict_resolution_by_merge_write(self):
        kv = DistributedKVStore(["s1", "s2", "s3"], n_replicas=3, w_quorum=2, r_quorum=2)

        base = kv.put("name", "john", coordinator="s1")
        c1 = kv.put("name", "johnSF", coordinator="s2", base_clock=base)
        c2 = kv.put("name", "johnNY", coordinator="s3", base_clock=base)

        merged_clock = dict(c1)
        for node, version in c2.items():
            merged_clock[node] = max(merged_clock.get(node, 0), version)

        kv.put("name", "johnMerged", coordinator="s1", base_clock=merged_clock)

        versions = kv.get("name")
        self.assertEqual(1, len(versions))
        self.assertEqual("johnMerged", versions[0].value)


if __name__ == "__main__":
    unittest.main()

File: key-value-store/implementation/cpp/test_kv_store.cpp
#include "kv_store.hpp"

#include <algorithm>
#include <cassert>
#include <iostream>
#include <string>
#include <vector>

using kvdemo::Clock;
using kvdemo::DistributedKVStore;

void test_put_get_single_value() {
    DistributedKVStore kv({"s1", "s2", "s3"}, 3, 2, 2);
    kv.put("user:1", "alice", "s1");
    auto versions = kv.get("user:1");
    assert(versions.size() == 1);
    assert(versions[0].value == "alice");
}

void test_concurrent_conflict() {
    DistributedKVStore kv({"s1", "s2", "s3"}, 3, 2, 2);
    Clock base = kv.put("name", "john", "s1");
    kv.put("name", "johnSF", "s2", base);
    kv.put("name", "johnNY", "s3", base);

    auto versions = kv.get("name");
    std::vector<std::string> vals;
    for (const auto& v : versions) vals.push_back(v.value);
    std::sort(vals.begin(), vals.end());

    assert(vals.size() == 2);
    assert(vals[0] == "johnNY");
    assert(vals[1] == "johnSF");
}

void test_merge_resolution() {
    DistributedKVStore kv({"s1", "s2", "s3"}, 3, 2, 2);
    Clock base = kv.put("name", "john", "s1");
    Clock c1 = kv.put("name", "johnSF", "s2", base);
    Clock c2 = kv.put("name", "johnNY", "s3", base);

    Clock merged = c1;
    for (const auto& [k, v] : c2) {
        if (!merged.count(k) || merged[k] < v) merged[k] = v;
    }

    kv.put("name", "johnMerged", "s1", merged);
    auto versions = kv.get("name");
    assert(versions.size() == 1);
    assert(versions[0].value == "johnMerged");
}

int main() {
    test_put_get_single_value();
    test_concurrent_conflict();
    test_merge_resolution();
    std::cout << "All C++ KV tests passed\n";
    return 0;
}
